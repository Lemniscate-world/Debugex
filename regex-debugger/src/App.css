:root {
  --primary-color: #4a6fa5;
  --secondary-color: #166088;
  --accent-color: #4fc3f7;
  --background-color: #f8f9fa;
  --text-color: #333;
  --error-color: #e57373;
  --success-color: #81c784;
  --border-radius: 8px;
  --box-shadow: 0 4px 6px rgba(0, 0, 0, 0.1);
  --transition: all 0.3s ease;
}

* {
  box-sizing: border-box;
  margin: 0;
  padding: 0;
}

body {
  font-family: 'Roboto', 'Segoe UI', Tahoma, Geneva, Verdana, sans-serif;
  background-color: var(--background-color);
  color: var(--text-color);
  line-height: 1.6;
}

#root {
  max-width: 1280px;
  margin: 0 auto;
  padding: 2rem;
}

.app {
  background-color: white;
  border-radius: var(--border-radius);
  box-shadow: var(--box-shadow);
  padding: 2rem;
  display: flex;
  flex-direction: column;
  gap: 1.5rem;
}

h1 {
  color: var(--primary-color);
  text-align: center;
  margin-bottom: 1.5rem;
  font-size: 2.2rem;
}

h2 {
  color: var(--secondary-color);
  margin-bottom: 0.8rem;
  font-size: 1.5rem;
}

.input-group {
  display: flex;
  flex-direction: column;
  gap: 0.5rem;
}

label {
  font-weight: 600;
  display: block;
  margin-bottom: 0.5rem;
}

input, textarea {
  width: 100%;
  padding: 0.8rem;
  border: 1px solid #ddd;
  border-radius: var(--border-radius);
  font-size: 1rem;
  transition: var(--transition);
}

input:focus, textarea:focus {
  outline: none;
  border-color: var(--accent-color);
  box-shadow: 0 0 0 2px rgba(79, 195, 247, 0.2);
}

textarea {
  min-height: 100px;
  resize: vertical;
}

.button-group {
  display: flex;
  gap: 1rem;
  justify-content: center;
  margin: 1rem 0;
}

button {
  background-color: var(--primary-color);
  color: white;
  border: none;
  border-radius: var(--border-radius);
  padding: 0.8rem 1.5rem;
  font-size: 1rem;
  cursor: pointer;
  transition: var(--transition);
}

button:hover {
  background-color: var(--secondary-color);
  transform: translateY(-2px);
}

button:disabled {
  background-color: #ccc;
  cursor: not-allowed;
  transform: none;
}

.execute-button {
  background-color: var(--accent-color);
  font-weight: bold;
  font-size: 1.1rem;
  padding: 1rem 2rem;
}

.execute-button:hover {
  background-color: #29b6f6;
}

.result-box {
  background-color: #f5f5f5;
  border-radius: var(--border-radius);
  padding: 1.5rem;
  margin-top: 0.5rem;
  border-left: 4px solid var(--primary-color);
}

.error-box {
  background-color: #ffebee;
  border-radius: var(--border-radius);
  padding: 1.5rem;
  margin-top: 0.5rem;
  border-left: 4px solid var(--error-color);
  color: #c62828;
}

.explanation-box {
  background-color: #e3f2fd;
  border-radius: var(--border-radius);
  padding: 1.5rem;
  margin-top: 0.5rem;
  border-left: 4px solid var(--accent-color);
}

.match-highlight {
  font-weight: bold;
  color: var(--secondary-color);
  background-color: rgba(79, 195, 247, 0.2);
  padding: 0.2rem 0.4rem;
  border-radius: 4px;
}

.sponsor-button {
  background-color: #ff5722;
  color: white;
  text-decoration: none;
  padding: 0.8rem 1.5rem;
  border-radius: var(--border-radius);
  display: inline-block;
  margin-top: 1rem;
  font-weight: bold;
  transition: var(--transition);
  text-align: center;
}

.sponsor-button:hover {
  background-color: #e64a19;
  transform: translateY(-2px);
}

.footer {
  text-align: center;
  margin-top: 2rem;
  color: #666;
  font-size: 0.9rem;
}

.step-indicator {
  display: flex;
  justify-content: center;
  align-items: center;
  margin-top: 0.5rem;
  font-size: 0.9rem;
  color: #666;
}
