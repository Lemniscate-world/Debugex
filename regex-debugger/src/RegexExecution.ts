// Interface for regex execution result
export interface RegexResult {
    matches: string[];
    error?: string;
}

// Function to execute regex step-by-step and highlight matches
export function executeRegexStepByStep(regex: string, text: string): RegexResult {
    const matches: string[] = [];
    let match;

    // Handle empty inputs
    if (!regex || !text) {
        return { matches, error: regex ? 'No text to match against' : 'No regex pattern provided' };
    }

    try {
        //https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/RegExp/RegExp
        const regexObj = new RegExp(regex, 'g'); // Global flag to find all matches

        while ((match = regexObj.exec(text)) !== null) {
            matches.push(match[0]);
        }

        return { matches };
    } catch (error) {
        return {
            matches,
            error: error instanceof Error ? error.message : 'Invalid regular expression'
        };
    }
}