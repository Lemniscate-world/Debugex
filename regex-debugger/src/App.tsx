import React, { useState } from 'react'
import { executeRegexStepByStep } from './RegexExecution';
import './App.css'
import { explainRegex } from './Explanation';
import { FaGithub } from 'react-icons/fa';

const App: React.FC = () => {
  const [regex, setRegex] = useState<string>('');
  const [text, setText] = useState<string>('');
  const [matches, setMatches] = useState<string[]>([]);
  const [currentStep, SetCurrentStep] = useState<number>(0);
  const [error, setError] = useState<string | undefined>();

  const handleExecute = () => {
    const result = executeRegexStepByStep(regex, text);
    setMatches(result.matches);
    setError(result.error);
    SetCurrentStep(0); // Set the current step to 0 when executing the regex
  }

  const handleNextStep = () => {
    if (currentStep < matches.length - 1) {
      SetCurrentStep(currentStep + 1);
    }
  };

  const handlePreviousStep = () => {
    if (currentStep > 0) {
      SetCurrentStep(currentStep - 1);
    }
  };

  const explanation = explainRegex(regex);

  return (
    <div className="app">
      <h1>Debugex - Regex Debugger + Learning Tool</h1>

      <div className="input-group">
        <label htmlFor="regex-input">Regex Pattern:</label>
        <input
          id="regex-input"
          type="text"
          value={regex}
          onChange={(e) => setRegex(e.target.value)}
          placeholder="Enter your regex pattern (e.g., \d+, [a-z]+)"
        />
      </div>

      <div className="input-group">
        <label htmlFor="text-input">Test Text:</label>
        <textarea
          id="text-input"
          value={text}
          onChange={(e) => setText(e.target.value)}
          placeholder="Enter text to match against your regex pattern"
        />
      </div>

      <button className="execute-button" onClick={handleExecute}>Execute Regex</button>

      <div className="button-group">
        <button onClick={handlePreviousStep} disabled={currentStep === 0 || matches.length === 0}>
          ← Previous Match
        </button>
        <button onClick={handleNextStep} disabled={currentStep === matches.length - 1 || matches.length === 0}>
          Next Match →
        </button>
      </div>

      {matches.length > 0 && (
        <div className="step-indicator">
          Match {currentStep + 1} of {matches.length}
        </div>
      )}

      <div>
        <h2>Explanation</h2>
        <div className="explanation-box">
          <p>{explanation || "Enter a regex pattern to see an explanation."}</p>
        </div>
      </div>

      <div>
        <h2>Matches</h2>
        {error ? (
          <div className="error-box">
            <p>Error: {error}</p>
          </div>
        ) : matches.length > 0 ? (
          <div className="result-box">
            <p>Current Match: <span className="match-highlight">{matches[currentStep]}</span></p>
          </div>
        ) : (
          <div className="result-box">
            <p>No matches found. Try a different pattern or text.</p>
          </div>
        )}
      </div>

      <div className="footer">
        <a
          href="https://ko-fi.com/kuroio"
          target="_blank"
          rel="noopener noreferrer"
          className="sponsor-button"
        >
          ❤️ Support on Ko-fi
        </a>
        <p style={{ marginTop: '1rem' }}>
          <a
            href="https://github.com/Lemniscate-world/debugex"
            target="_blank"
            rel="noopener noreferrer"
            style={{ color: 'var(--primary-color)', textDecoration: 'none', display: 'flex', alignItems: 'center', justifyContent: 'center', gap: '0.5rem' }}
          >
            <FaGithub /> View on GitHub
          </a>
        </p>
      </div>
    </div>
  );
}

export default App
