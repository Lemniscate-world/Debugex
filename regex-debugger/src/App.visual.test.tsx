import { describe, it, expect, vi, beforeEach } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import App from './App';
import * as RegexExecution from './RegexExecution';

// Mock the RegexExecution module
vi.mock('./RegexExecution', () => ({
  executeRegexStepByStep: vi.fn()
}));

describe('App Visual Regression Tests', () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it('renders the initial UI correctly', () => {
    const { container } = render(<App />);
    
    // Check that all main UI elements are present
    expect(screen.getByText('Regex Debugger + Learning Tool')).toBeInTheDocument();
    expect(screen.getByText('Regex:')).toBeInTheDocument();
    expect(screen.getByText('Text:')).toBeInTheDocument();
    expect(screen.getByText('Execute')).toBeInTheDocument();
    expect(screen.getByText('Previous Step')).toBeInTheDocument();
    expect(screen.getByText('Next Step')).toBeInTheDocument();
    expect(screen.getByText('Explanation:')).toBeInTheDocument();
    expect(screen.getByText('Matches:')).toBeInTheDocument();
    
    // Initial state should show "No matches found"
    expect(screen.getByText('No matches found.')).toBeInTheDocument();
    
    // Snapshot the initial UI
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders matches correctly', () => {
    const mockExecuteRegex = vi.mocked(RegexExecution.executeRegexStepByStep);
    mockExecuteRegex.mockReturnValue({ matches: ['match1', 'match2'] });
    
    const { container } = render(<App />);
    
    // Execute the regex
    fireEvent.click(screen.getByText('Execute'));
    
    // Check that the match is displayed
    expect(screen.getByText('Current Match: match1')).toBeInTheDocument();
    
    // Snapshot the UI with matches
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders error state correctly', () => {
    const mockExecuteRegex = vi.mocked(RegexExecution.executeRegexStepByStep);
    mockExecuteRegex.mockReturnValue({ matches: [], error: 'Invalid regular expression' });
    
    const { container } = render(<App />);
    
    // Execute the regex
    fireEvent.click(screen.getByText('Execute'));
    
    // Check that the error is displayed
    expect(screen.getByText('Error: Invalid regular expression')).toBeInTheDocument();
    
    // Snapshot the UI with error
    expect(container.innerHTML).toMatchSnapshot();
  });

  it('renders step navigation correctly', () => {
    const mockExecuteRegex = vi.mocked(RegexExecution.executeRegexStepByStep);
    mockExecuteRegex.mockReturnValue({ matches: ['match1', 'match2', 'match3'] });
    
    const { container } = render(<App />);
    
    // Execute the regex
    fireEvent.click(screen.getByText('Execute'));
    
    // Check initial state
    expect(screen.getByText('Current Match: match1')).toBeInTheDocument();
    
    // Navigate to next step
    fireEvent.click(screen.getByText('Next Step'));
    expect(screen.getByText('Current Match: match2')).toBeInTheDocument();
    
    // Snapshot after navigation
    expect(container.innerHTML).toMatchSnapshot();
    
    // Navigate to next step again
    fireEvent.click(screen.getByText('Next Step'));
    expect(screen.getByText('Current Match: match3')).toBeInTheDocument();
    
    // Previous button should be enabled, next button should be disabled
    const prevButton = screen.getByText('Previous Step');
    const nextButton = screen.getByText('Next Step');
    expect(prevButton).not.toBeDisabled();
    expect(nextButton).toBeDisabled();
    
    // Navigate back
    fireEvent.click(prevButton);
    expect(screen.getByText('Current Match: match2')).toBeInTheDocument();
    
    // Final snapshot
    expect(container.innerHTML).toMatchSnapshot();
  });
});
