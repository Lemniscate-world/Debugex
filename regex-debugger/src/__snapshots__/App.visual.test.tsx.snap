// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`App Visual Regression Tests > renders error state correctly 1`] = `"<div class="app"><h1>Regex Debugger + Learning Tool</h1><div><label>Regex:<input placeholder="Enter your regex" type="text" value=""></label></div><div><label>Text:<textarea placeholder="Enter text to match against"></textarea></label></div><button>Execute</button><div><button disabled="">Previous Step</button><button>Next Step</button></div><div><h2>Explanation:</h2><p>No explanation available for this pattern</p></div><div><h2>Matches:</h2><div style="background-color: rgb(255, 221, 221); padding: 10px; color: red;"><p>Error: Invalid regular expression</p></div></div></div>"`;

exports[`App Visual Regression Tests > renders matches correctly 1`] = `"<div class="app"><h1>Regex Debugger + Learning Tool</h1><div><label>Regex:<input placeholder="Enter your regex" type="text" value=""></label></div><div><label>Text:<textarea placeholder="Enter text to match against"></textarea></label></div><button>Execute</button><div><button disabled="">Previous Step</button><button>Next Step</button></div><div><h2>Explanation:</h2><p>No explanation available for this pattern</p></div><div><h2>Matches:</h2><div style="background-color: rgb(244, 244, 244); padding: 10px;"><p>Current Match: match1</p></div></div></div>"`;

exports[`App Visual Regression Tests > renders step navigation correctly 1`] = `"<div class="app"><h1>Regex Debugger + Learning Tool</h1><div><label>Regex:<input placeholder="Enter your regex" type="text" value=""></label></div><div><label>Text:<textarea placeholder="Enter text to match against"></textarea></label></div><button>Execute</button><div><button>Previous Step</button><button>Next Step</button></div><div><h2>Explanation:</h2><p>No explanation available for this pattern</p></div><div><h2>Matches:</h2><div style="background-color: rgb(244, 244, 244); padding: 10px;"><p>Current Match: match2</p></div></div></div>"`;

exports[`App Visual Regression Tests > renders step navigation correctly 2`] = `"<div class="app"><h1>Regex Debugger + Learning Tool</h1><div><label>Regex:<input placeholder="Enter your regex" type="text" value=""></label></div><div><label>Text:<textarea placeholder="Enter text to match against"></textarea></label></div><button>Execute</button><div><button>Previous Step</button><button>Next Step</button></div><div><h2>Explanation:</h2><p>No explanation available for this pattern</p></div><div><h2>Matches:</h2><div style="background-color: rgb(244, 244, 244); padding: 10px;"><p>Current Match: match2</p></div></div></div>"`;

exports[`App Visual Regression Tests > renders the initial UI correctly 1`] = `"<div class="app"><h1>Regex Debugger + Learning Tool</h1><div><label>Regex:<input placeholder="Enter your regex" type="text" value=""></label></div><div><label>Text:<textarea placeholder="Enter text to match against"></textarea></label></div><button>Execute</button><div><button disabled="">Previous Step</button><button>Next Step</button></div><div><h2>Explanation:</h2><p>No explanation available for this pattern</p></div><div><h2>Matches:</h2><p>No matches found.</p></div></div>"`;
